cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,240 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// returnGegarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <utility>
+
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernels/take.h"
+#include "arrow/util/logging.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+namespace compute {
+
+Status Take(FunctionContext* context, const Array& values, const Array& indices,
+            const TakeOptions& options, std::shared_ptr<Array>* out) {
+  Datum out_datum;
+  RETURN_NOT_OK(
+      Take(context, Datum(values.data()), Datum(indices.data()), options, &out_datum));
+  *out = out_datum.make_array();
+  return Status::OK();
+}
+
+Status Take(FunctionContext* context, const Datum& values, const Datum& indices,
+            const TakeOptions& options, Datum* out) {
+  TakeKernel kernel(values.type(), options);
+  RETURN_NOT_OK(kernel.Call(context, values, indices, out));
+  return Status::OK();
+}
+
+struct TakeParameters {
+  FunctionContext* context;
+  std::shared_ptr<Array> values, indices;
+  TakeOptions options;
+  std::shared_ptr<Array>* out;
+};
+
+template <typename Builder, typename Scalar>
+Status UnsafeAppend(Builder* builder, Scalar&& value) {
+  builder->UnsafeAppend(std::forward<Scalar>(value));
+  return Status::OK();
+}
+
+Status UnsafeAppend(BinaryBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
+  builder->UnsafeAppend(value);
+  return Status::OK();
+}
+
+Status UnsafeAppend(StringBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
+  builder->UnsafeAppend(value);
+  return Status::OK();
+}
+
+template <TakeOptions::OutOfBoundsBehavior B, bool AllValuesValid, bool AllIndicesValid,
+          typename ValueArray, typename IndexArray, typename OutBuilder>
```

Comment:
Looks like this is going to generate a large number of functions... This would be a good reason to remove `OutOfBoundsBehavior` and only implement RAISE.